scripts/install/install_unix.sh

```diff
@@ -13,165 +13,638 @@
+    fi
+
+  elif [ -n "$http_proxy" ]; then
+    info "Using proxy from profile: $http_proxy"
+    proxy="$http_proxy"
+  elif [ -n "$https_proxy" ]; then
+    info "Using proxy from profile: $https_proxy"
+    proxy="$https_proxy"
+  fi
+
+  if [ -z "$full_proxy" ]; then
+    full_proxy="$proxy"
+  fi
+}
+
+# This will set the os based on the current runtime environment.
+# Accepted values are darwin and linux. This value cannot be overriden.
+set_os()
+{
+  os_key=$(uname -s)
+  case "$os_key" in
+    Linux)
+      os="linux"
+      ;;
+    *)
+      error "Unsupported os type: $os_key"
+      ;;
+  esac
+}
 
-set_arch() {
-    os_arch=$(uname -m)
-    case "$os_arch" in
+
+set_os_arch()
+{
+  os_arch=$(uname -m)
+  case "$os_arch" in 
     # arm64 strings. These are from https://stackoverflow.com/questions/45125516/possible-values-for-uname-m
     aarch64|arm64|aarch64_be|armv8b|armv8l)
-        arch="arm64"
-        ;;
+      os_arch="arm64"
+      ;;
     x86_64)
-        arch="amd64"
-        ;;
+      os_arch="amd64"
+      ;;
     # armv6/32bit. These are what raspberry pi can return, which is the main reason we support 32-bit arm
     arm|armv6l|armv7l)
-        arch="arm"
-        ;;
+      os_arch="arm"
+      ;;
     *)
-        echo "Installation failed: Unsupported os arch: $os_arch"
-        exit 1
-        ;;
-    esac
+      error "Unsupported os arch: $os_arch"
+      ;;
+  esac
+}
+
+# Set the package type before install
+set_package_type()
+{
+  if command -v dpkg > /dev/null 2>&1; then
+      package_type="deb"
+  elif command -v rpm > /dev/null 2>&1; then
+      package_type="rpm"
+  else
+      error "Could not find dpkg or rpm on the system"
+  fi
+}
+
+# This will set the urls to use when downloading the collector and its plugins.
+# These urls are constructed based on the --version flag or COLLECTOR_VERSION env variable.
+# If not specified, the version defaults to "latest".
+set_download_urls()
+{
+  if [ -z "$version" ] ; then
+    # shellcheck disable=SC2153
+    version=$COLLECTOR_VERSION
+  fi
+
+  if [ -z "$url" ] ; then
+    url=$DOWNLOAD_BASE
+  fi
+
+  if [ -z "$arch" ] ; then 
+    os_arch=$arch
+  fi
+
+  if [ -z "$pkg_type" ] ; then
+    package_type=$pkg_type
+  fi
+
+  if [ -z "$version" ] ; then
+    collector_download_url="$url/latest/download/${PACKAGE_NAME}_${os_arch}.${package_type}"
+  else
+    collector_download_url="$url/download/v$version/${PACKAGE_NAME}_${os_arch}.${package_type}"
+  fi
+}
+
+# This will check all prerequisites before running an installation.
+check_prereqs()
+{
+  banner "Checking Prerequisites"
+  increase_indent
+  root_check
+  os_check
+  os_arch_check
+  package_type_check
+  dependencies_check
+  success "Prerequisite check complete!"
+  decrease_indent
+}
+
+# This
+root_check()
+{
+  system_user_name=$(id -un)
+  if [ "${system_user_name}" != 'root' ]
+  then
+    failed
+    error_exit "Script needs to be run as root or with sudo permissions"
+  fi
+}
```

Comment:
Comment here was left unfinished;

I'm wondering if this is the best way to handle checking for root. There seem to be pitfalls here that might not work everywhere. https://stackoverflow.com/questions/18215973/how-to-check-if-running-as-root-in-a-bash-script

Is it better to check if `id -u` is 0? I know this is super edgecase, but it's something I worry about.

Also, the error message indicates it should be run "with sudo permissions", but I would change this to be "with sudo", since you need more that just the permissions, the script does not invoke sudo for you.

Reply:
I'll fixup the comment and error message.

I actually tried what that stackoverflow link suggested and it didn't work. I think this is fine it's an edge case.

Reply:
The comment is still left empty here

Reply:
Good catch. I think I got distracted by the other points.